state/protocol/badger/mutator.go

```diff
@@ -770,12 +771,12 @@ func (m *FollowerState) isFirstBlockOfEpoch(block *flow.Header, currentEpochSetu
 	currentEpochFirstView := currentEpochSetup.FirstView
 	// sanity check: B.View >= W
 	if block.View < currentEpochFirstView {
-		return false, fmt.Errorf("[unexpected] data inconsistency: block (id=%x, view=%d) is below its epoch first view %d", block.ID(), block.View, currentEpochFirstView)
+		return false, irrecoverable.NewExceptionf("data inconsistency: block (id=%x, view=%d) is below its epoch first view %d", block.ID(), block.View, currentEpochFirstView)
 	}
 
 	parent, err := m.headers.ByBlockID(block.ParentID)
 	if err != nil {
-		return false, fmt.Errorf("[unexpected] could not retrieve parent (id=%s): %v", block.ParentID, err)
+		return false, irrecoverable.NewExceptionf("could not retrieve parent (id=%s): %w", block.ParentID, err)
```

Comment:
This is exactly why I think we _should_ liberally wrap with `NewExceptionf`. `ByBlockID` might return a generic storage error (exception) or `ErrNotFound`. `isFirstBlockOfEpoch` is the level at which we can interpret `ErrNotFound` as an exception, because contextually we know that the parent block must be stored.

> We could equally well just wrap higher up, when there is a potential for conflicts (?)

The problem is that the point at which there might be conflicts could be arbitrarily many levels further up the stack. 

We document that no errors are expected from `isFirstBlockOfEpoch` so the _caller_ knows that any error is any exception. If the stack was only 2 frames deep we would be fine. However, suppose the stack is several functions deep. Several frames above `isFirstBlockOfEpoch` we might be calling some function `f` which _does_ document `ErrNotFound` as an expected sentinel. Unless the sentinel is wrapped _at the point it is interpreted as an exception_, it will be misinterpreted by the caller of `f` as an expected sentinel. 

Nowhere between `f` and `isFirstBlockOfEpoch` would `ErrNotFound` be documented, so I think it is unfeasible to expect that this could be handled at the point where there might be conflicts.